src/pyo3_extensions/chrono_tz.rs

```diff
@@ -0,0 +1,232 @@
+            .expect("Failed to construct datetime");
+        datetime.into()
+    }
+}
+
+impl IntoPy<PyObject> for ChronoDateTime {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        ToPyObject::to_object(&self, py)
+    }
+}
+
+impl FromPyObject<'_> for ChronoDateTime {
+    fn extract(ob: &PyAny) -> PyResult<ChronoDateTime> {
+        let mut dt = <PyDateTime as PyTryFrom>::try_from(ob)?;
+        let tz: chrono_tz::Tz = if let Some(tzinfo) = dt.get_tzinfo() {
+            // If we have a "key" (name) for the timezone, we try to parse it with
+            // chrono_tz. Otherwise we ask python to convert the datetime to UTC
+            // and return UTC as the timezone.
+            if let Ok(key) = tzinfo.getattr("key") {
+                let key: String = key.extract()?;
+                key.parse()
+                    .map_err(|_err| PyTypeError::new_err("Timezone not found"))?
+            } else {
+                let utc = ob
+                    .py()
+                    .import("datetime")?
+                    .getattr("timezone")?
+                    .getattr("utc")?;
+                dt = dt.getattr("astimezone")?.call1((utc,))?.cast_as()?;
+                chrono_tz::Tz::UTC
+            }
+        } else {
+            // If we receive a naivedatetime, we ask python to convert it to UTC.
+            // Python assumes that naive datetimes are in the local timezone, so this
+            // is not like assuming it was in UTC.
+            // https://docs.python.org/3/library/datetime.html#datetime.datetime.astimezone
+            // TODO: Should we instead assume utc, and not convert? Here we are converting
+            //       to the local timezone of the system running the dataflow, but the date
+            //       could come from a different place, so whatever we do is probably wrong.
+            warn!("Received naive datetime! Converting to utc assuming local datetime");
+            let utc = ob
+                .py()
+                .import("datetime")?
+                .getattr("timezone")?
+                .getattr("utc")?;
+            dt = dt.getattr("astimezone")?.call1((utc,))?.cast_as()?;
+            chrono_tz::Tz::UTC
+        };
+
+        let ms = dt.get_fold() as u32 * 1_000_000 + dt.get_microsecond();
+        let h = dt.get_hour().into();
+        let m = dt.get_minute().into();
+        let s = dt.get_second().into();
+        let dt = NaiveDateTime::new(
+            NaiveDate::from_ymd(dt.get_year(), dt.get_month().into(), dt.get_day().into()),
+            NaiveTime::from_hms_micro(h, m, s, ms),
+        );
+        if let Some(tz_dt) = tz.from_local_datetime(&dt).earliest() {
+            Ok(ChronoDateTime(tz_dt))
+        } else {
+            Err(PyTypeError::new_err("Can't convert datetime"))
+        }
+    }
+}
+
+// Implement serde (de)serialization manually here, since there is no standard
+// way to serialize datetimes with IANA timezone info (yet?).
```

Comment:
Your understanding is correct, the point of using tz rather than a fixed offset is to keep the same information if the user provided a timezone info rather than an offset, see answers below for more context.